cassiopeia/plotting/itol_utilities.py

```diff
@@ -0,0 +1,623 @@
+    cassiopeia_tree: CassiopeiaTree,
+    tree_name: str,
+    export_filepath: str,
+    itol_config: str = "~/.itolconfig",
+    api_key: Optional[str] = None,
+    project_name: Optional[str] = None,
+    meta_data: List[str] = [],
+    allele_table: Optional[pd.DataFrame] = None,
+    indel_colors: Optional[pd.DataFrame] = None,
+    indel_priors: Optional[pd.DataFrame] = None,
+    rect: bool = False,
+    include_legend: bool = False,
+    palette: List[str] = palettes.Category20[20],
+    random_state: Optional[np.random.RandomState] = None,
+):
+    """Uploads a tree to iTOL and exports it.
+
+    This function takes in a tree, plots in with iTOL, and exports it locally.
+    A user can also specify meta data and allele tables to visualize alongside
+    their tree. The function requires a user to have an account with iTOL and
+    must pass in an `api_key` and a `project_name`, which corresponds to one
+    in the user's iTOL account.
+
+    TODO(mgjones): Add the ability to pass in min/max colors for specific
+        numeric meta data to use when creating the gradient files.
+
+    Args:
+        cassiopeia_tree: A CassiopeiaTree instance, populated with a tree.
+        api_key: API key linking to your iTOL account
+        project_name: Project name to upload to.
+        tree_name: Name of the tree. This is what the tree will be called
+            within your project directory on iTOL
+        export_filepath: Output file path to save your tree. Must end with
+            one of the following suffixes: ['png', 'svg', 'eps', 'ps', 'pdf'].
+        meta_data: Meta data to plot alongside the tree, which must be columns
+            in the CassiopeiaTree.cell_meta variable.
+        allele_table: Alleletable to plot alongside the tree.
+        indel_colors: Color mapping to use for plotting the alleles for each
+            cell. Only necessary if `allele_table` is specified.
+        indel_priors: Prior probabilities for each indel. Only useful if an
+            allele table is to be plotted and `indel_colors` is None.           
+        rect: Boolean indicating whether or not to save your tree as a circle
+            or rectangle.
+        include_legend: Plot legend along with meta data.
+        palette: A palette of colors in hex format.
+        random_state: A random state for reproducibility
+    """
+
+    # create temporary directory for storing files we'll upload to iTOL
+    temporary_directory = tempfile.mkdtemp()
+
+    if os.path.exists(os.path.expanduser(itol_config)):
+        
+        config = configparser.ConfigParser()
+        with open(os.path.expanduser(itol_config), "r") as f:
+            config_string = f.read()
+        config.read_string(config_string)
+        if api_key is None:
+            api_key = config["DEFAULT"]["api_key"]
+        if project_name is None:
+            project_name = config["DEFAULT"]["project_name"]
+
+    if api_key is None or project_name is None:
```

Comment:
We can be a bit more conservative here and say: `if itol_config is None and (api_key is None or project_name is None)`; the current `if` condition relies on previous logic, which can potentially lead to bugs if the previous code changed.

Reply:
Good point. I've changed up the logic here to take this into account. Hope it makes sense.